### PR TITLE
proof-of-concept for creating an archivy CLI

### PR DESCRIPTION
**Pull request by clemux**
Related: #108
There a few issues I haven't managed to fix yet.
First, I need to `export FLASK_APP=archivy`. Second, I had to comment out those two lines in __init__.py:

```python
if 'pytest' not in sys.argv[0]:
    Thread(target=run_watcher, args=[app]).start()
```

Otherwise, the thread won't die and the CLI won't exist.

```
(venv) [clemux@nardole archivy]$ archivy setup
Setting up archivy...
(venv) [clemux@nardole archivy]$ archivy
Usage: archivy [OPTIONS] COMMAND [ARGS]...

Options:
  --version  Show the flask version
  --help     Show this message and exit.

Commands:
  routes  Show the routes for the app.
  run
  setup
  shell   Run a shell in the app context.
(venv) [clemux@nardole archivy]$ archivy routes
Endpoint             Methods      Rule
-------------------  -----------  ------------------------------------
api.change_bookmark  PUT          /api/bookmarks/<int:bookmark_id>
api.create_bookmark  POST         /api/bookmarks
api.create_note      POST         /api/notes
api.delete_bookmark  DELETE       /api/dataobjs/<int:dataobj_id>
api.get_dataobj      GET          /api/dataobjs/<int:dataobj_id>
api.get_dataobjs     GET          /api/dataobjs
api.local_edit       GET          /api/dataobj/local_edit/<dataobj_id>
create_folder        POST         /folders/new
delete_data          DELETE, GET  /dataobj/delete/<dataobj_id>
delete_folder        DELETE       /folders/delete
index                GET          /index
index                GET          /
new_bookmark         GET, POST    /bookmarks/new
new_note             GET, POST    /notes/new
parse_pocket         GET          /parse_pocket
pocket_settings      GET, POST    /pocket
search_elastic       GET          /search
show_dataobj         GET          /dataobj/<dataobj_id>
static               GET          /static/<path:filename>
(venv) [clemux@nardole archivy]$ archivy setup
Setting up archivy...
(venv) [clemux@nardole archivy]$ archivy run
Running archivy...
/home/clemux/Projects/archivy/archivy/cli.py:19: Warning: Silently ignoring app.run() because the application is run from the flask command line executable.  Consider putting app.run() behind an if __name__ == "__main__" guard to silence this warning.
  app.run(host='0.0.0.0', port=port)
(venv) [clemux@nardole archivy]$ 
```

Still, that's a start and shows that `click` is awesome ;)

### CODE REVIEW

**Comment by Uzay-G** [comment]
Will check it out this weekend!

**Comment by Uzay-G** [comment]
I fixed the thread problem by initialising the thread only inside the `run` command instead of the `__init__` module.

I also followed the instructions https://flask.palletsprojects.com/en/1.1.x/cli/#custom-scripts here so that we don't need to set the `FLASK_APP`.

**Comment by Uzay-G** [comment]
Once we've tested this, I'll merge it in :+1:. @srevinsaju could you maybe check this out if you have time? I just want to make sure everything works correctly on someone else's machine.

**Comment by srevinsaju** [comment]
Sure. I will test it soon.

**Comment by srevinsaju** [comment]
Reviewed and tested. I would consider not merging this, (preferably), until the following are fixed:
* Requires you to press Ctrl+C twice to quit archivy when using `archivy run`. This is a bug
* Second Ctrl+C raises an error, but instead, it should have quit silently?
* the function, `run_watcher` should be converted into a class, and include proper `stop` method which would terminate that thread, instead of keeping it running, as a potential memory leak can occur if we have an exception while runnign archivy, i.e not by a `KeyboardInterrupt`
* preferably, until the `setup` function is added, we could remove the `setup` command, to prevent confusing poor users.

Stuff which are not under the jurisdiction of this PR:
* does not check if `ARCHIVY_PORT` is a valid integer. Instead of raising `ValueError`, it should raise `archivy.PortConfigurationError`. because I can set `ARCHIVY_PORT=abcd`, which would raise a `ValueError` which should be meaningfully translated to `archivy.PortConfigurationError`

**Comment by Uzay-G** [comment]
the pytest check is no longer needed, and I think with @clemux's help we will try to add the config thing before the merge. I will look into the other important issues you mentioned when I have time.

File: archivy/check_changes.py [hunk 1]
```diff
@@ -85,21 +86,29 @@ def on_created(self, event):
...
+            ),
+            recursive=True)
+        observer.start()
+
+        while self.running:
             time.sleep(1)
-    except KeyboardInterrupt:
         observer.stop()
-    observer.join()
+        observer.join()
```
Comment (srevinsaju) [inline archivy/check_changes.py:111]: I guess `observer.join` should be called before `observer.stop`

File: archivy/cli.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    load_dotenv()
+    # prevent pytest from hanging because of running thread
+    watcher = Watcher(app)
+    watcher.start()
+    port = int(os.environ.get("ARCHIVY_PORT", 5000))
+    os.environ["FLASK_RUN_FROM_CLI"] = "false"
+    app.run(host='0.0.0.0', port=port)
+    # this code will be reached when the user stops running the server by doing CTRL+C
+    watcher.stop()
+    watcher.join()
```
Comment (srevinsaju) [inline archivy/cli.py:31]: Same here, probably `join` should come before `stop`.

File: archivy/cli.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+def run():
+    click.echo('Running archivy...')
+    load_dotenv()
+    # prevent pytest from hanging because of running thread
+    watcher = Watcher(app)
+    watcher.start()
+    port = int(os.environ.get("ARCHIVY_PORT", 5000))
+    os.environ["FLASK_RUN_FROM_CLI"] = "false"
+    app.run(host='0.0.0.0', port=port)
+    # this code will be reached when the user stops running the server by doing CTRL+C
```
Comment (srevinsaju) [inline archivy/cli.py:29]: I hope `try..except` block should fix the double `Ctrl+C`? Did you try that one?

**Review by srevinsaju — CHANGES_REQUESTED** [review]

File: archivy/check_changes.py (hunk 1 above)
Comment (Uzay-G) [reply archivy/check_changes.py:111]: I'm taking from the lib's example: https://python-watchdog.readthedocs.io/en/v0.10.2/quickstart.html#a-simple-example

File: archivy/cli.py (hunk 2 above)
Comment (Uzay-G) [reply archivy/cli.py:31]: I think it's supposed to be like that where you stop the thread and then let it collect itself.

File: archivy/check_changes.py (hunk 1 above)
Comment (srevinsaju) [reply archivy/check_changes.py:111]: Yes, but that example has `.join` inside the `try:` block. 
So the minimum or basic code flow is 
`t.join() -> t.stop() -> t.join()`

But in this case, its this
`t.stop() -> t.join() `

Comment (srevinsaju) [reply archivy/check_changes.py:111]: Ok, nvm. I am wrong. You are right :sweat_smile:

File: archivy/cli.py (hunk 2 above)
Comment (srevinsaju) [reply archivy/cli.py:31]: Agreed. Please mark as resolved

File: archivy/check_changes.py (hunk 1 above)
Comment (srevinsaju) [reply archivy/check_changes.py:111]: I wonder where the `Mark as resolved` button went. anyway, its resolved :rocket:

Comment (Uzay-G) [reply archivy/check_changes.py:111]: So should it become `t.join()` -> `t.stop()`?

**Comment by srevinsaju** [comment]
Tested now. It works with a single Ctrl+C, but the Ctrl+C takes something like 2 seconds.

```diff
diff --git a/archivy/cli.py b/archivy/cli.py
index e9e613f..1e1fad6 100644
--- a/archivy/cli.py
+++ b/archivy/cli.py
@@ -27,8 +27,10 @@ def run():
     os.environ["FLASK_RUN_FROM_CLI"] = "false"
     app.run(host='0.0.0.0', port=port)
     # this code will be reached when the user stops running the server by doing CTRL+C
+    click.echo("Stopping archivy.watcher")
     watcher.stop()
     watcher.join()
+    click.echo("Bye!")
```

this would make it more responsive, so that it does not make the user feel like hitting Ctrl + C for a second time,

**Comment by srevinsaju** [comment]
Please defer this merge, `flask.cli.run` and `archivy.cli.run` conflicts. 
Proof:
Rename `run` to `foobar` in `archivy/cli.py` and try running this on terminal
```bash
archivy run
```

The server starts, and is the right way of doing this. If you would like to do some other function before the server starts, preferably:
* Create `ArchivyFlaskGroup` inherits `FlaskGroup`  and modify accordingly
* Remove `FlaskGroup` imports and manually write the Flask based functions

**Comment by Uzay-G** [comment]
I fixed it by removing the default `run_command` as suggested [here](https://github.com/pallets/flask/issues/2020). What do you think?

**Review by srevinsaju — APPROVED** [review]
https://github.com/Uzay-G/archivy/pull/109/commits/3620163763bec32b0af8cb42e7ddc7d91727c2a1 fixed it. Great! Tested and reviewed. Works fine. Ready to merge :tada:

**Comment by Uzay-G** [comment]
Cool!

**Pull request merged**